spec/abstract/SchedContext_A.thy

```diff
@@ -304,13 +304,12 @@ where
      usage' \<leftarrow> handle_overrun_loop usage;
 
      refills \<leftarrow> get_refills sc_ptr;
-     new_head_amount \<leftarrow> return (r_amount (hd refills));
 
      when (usage' > 0 \<and> r_time (hd refills) < MAX_RELEASE_TIME) $ do
        sc \<leftarrow> get_sched_context sc_ptr;
-       period \<leftarrow> return (sc_period sc);
-       used \<leftarrow> return \<lparr>r_time = r_time (hd (sc_refills sc)) + period, r_amount = usage'\<rparr>;
-       update_refill_hd sc_ptr (r_time_update (\<lambda>t. t + usage') o (r_amount_update (\<lambda>m. m - usage')));
+       used \<leftarrow> return \<lparr>r_time = r_time (hd (sc_refills sc)) + sc_period sc, r_amount = usage'\<rparr>;
+       set_refill_hd sc_ptr \<lparr>r_time = r_time (refill_hd sc) + usage',
+                             r_amount = r_amount (refill_hd sc) - usage'\<rparr>;
```

Comment:
I am not too sure of this change, to be honest. What is the reason behind it? If it is just aligning with Haskell, wouldn’t it be better to change Haskell to be similar to the abstract for this case, if you consider how the C code is?

Reply:
It was to ease the refinement. If you're happy for me to change this part of `refillBudgetCheck`, I'll get rid of the commits related to the change to the definition of `refill_budget_check` and will use two calls to `update_refill_hd/updateRefillHd` in `refill_budget_check/refillBudgetCheck`

Reply:
I suppose I could just leave the asbtract spec as it is and then use a rewrite rule for the `corres` rule. I might just go ahead with that

Reply:
It would be helpful for later to change Haskell to use `updateRefillHd` instead of `setRefillHd` while we are here, wouldn’t it?

Reply:
Yeah, I'd like to leave the definition of `refill_budget_check` as it is in current `rt`, replace the call to `setRefillHd` in `refillBudgetCheck` with two calls to `updateRefillHd`, and then use a rewrite rule to help with the proof of `refillBudgetCheck_corres`

Reply:
Great, that’ll be perfect :)

Reply:
The changes are here

https://github.com/michaelmcinerney/l4v/commits/refillBudgetCheck_corres3

I squashed the commits relevant to the `valid_refills` conjunct and the generalised helper lemmas into the first commit in that branch. I can force push that branch to this branch if that's better.

I think the best way forward would be to wait for the `refillUnblockCheck_corres` PR to be merged, and then I can rebase the linked branch over that, move the work in `SchedContextInv_R.thy` to `Schedule_R.thy`, and then update the guards in the new `corres` rules I have